### PR TITLE
haskell.compiler: allow overriding source with hadrian

### PR DESCRIPTION
**Pull request by 9999years**
Hadrian (the GHC build tool) is built separately from GHC. This means that if `haskell.compiler.ghc961` is overridden to add patches, those patches will _only_ be applied to the GHC portion of the build, and not the Hadrian build. For example, backporting this patch to GHC 9.6.1 failed because the changes to `hadrian/` files were not reflected in the Nix build:

[Enable response files for linker if supported](https://gitlab.haskell.org/ghc/ghc/-/commit/5ed77deb1b49bff7bae9660487ee4a40a496476d)

By lifting `src` and `hadrian` from variables defined in the function body to parameters with default values, the `hadrian/` files can be overridden using the `haskell.compiler.ghc961.override` function. For example:

```nix
self.haskell.compiler.ghc961.override {
  # The GHC 9.6 builder in nixpkgs first builds hadrian with the
  # source tree provided here and then uses the built hadrian to
  # build the rest of GHC. We need to make sure our patches get
  # included in this `src`, then, rather than modifying the tree in
  # the `patchPhase` or `postPatch` of the outer builder.
  src = self.applyPatches {
    src = let
      version = "9.6.1";
    in
      self.fetchurl {
        url = "https://downloads.haskell.org/ghc/${version}/ghc-${version}-src.tar.xz";
        sha256 = "fe5ac909cb8bb087e235de97fa63aff47a8ae650efaa37a2140f4780e21f34cb";
      };

    patches = [
      # Enable response files for linker if supported
      (self.fetchpatch {
        url = "https://gitlab.haskell.org/ghc/ghc/-/commit/5ed77deb1b49bff7bae9660487ee4a40a496476d.patch";
        hash = "sha256-dvenK+EPTZJYXnyfKPdkvLp+zeUmsY9YrWpcGCzYStM=";
      })
    ];
  };
}
```

Note that we do have to re-declare the `src` we want, but I'm not sure of a good way to avoid this while also sharing one set of patches between the GHC and Hadrian builds.

### CODE REVIEW

File: pkgs/development/compilers/ghc/common-hadrian.nix [hunk 1]
```diff
@@ -161,13 +175,7 @@
...
 
 let
-  src = (if rev != null then fetchgit else fetchurl) ({
-    inherit url sha256;
-  } // lib.optionalAttrs (rev != null) {
-    inherit rev;
-  });
-
-  inherit (stdenv) buildPlatform hostPlatform targetPlatform;
+  inherit (stdenv) hostPlatform targetPlatform;
```
Comment (sternenseemann) [inline pkgs/development/compilers/ghc/common-hadrian.nix:178]: Unnecessary to change.

```suggestion
  inherit (stdenv) buildPlatform hostPlatform targetPlatform;
```

**Review by sternenseemann — APPROVED** [review]
Very nice!

**Comment by sternenseemann** [comment]
I suppose an alternative would be to pass `finalAttrs.src` and `finalAttrs.patches` to hadrian (and start using `finalAttrs`), but I'm not sure if that's better?

The big problem with hadrian is that it would be cleanest to have hadrian be an ordinary package in `bootPkgs`, but since `bootPkgs` can't know what GHC it'll build (in fact it can build multiple GHCs), this doesn't seem to be possible and we have this annoying inline `callPackage`. Maybe we should in the future have a per-compiler bootstrap package set that can be overridden. In any case, your solution seems more convenient for now.

**Review by sternenseemann — CHANGES_REQUESTED** [review]
Okay the argument default does not work sadly, since we have `pkgs.src` which is passed in by `callPackage`. Not sure how to best fix this at the moment…

```
unpacking sources
unpacking source archive /nix/store/56321rcxigd7ad6q1i7lixkdq3804y1f-src-1.31
source root is src-1.31
find: ‘src-1.31/hadrian’: No such file or directory
chmod: cannot access 'src-1.31/hadrian': No such file or directory
```

File: pkgs/development/compilers/ghc/common-hadrian.nix (hunk 1 above)
Comment (9999years) [reply pkgs/development/compilers/ghc/common-hadrian.nix:178]: Why? `buildPlatform` isn't used for anything in this file.

**Comment by 9999years** [comment]
I renamed it to `ghcSrc`, like the `hadrian` package uses.

File: pkgs/development/compilers/ghc/common-hadrian.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/compilers/ghc/common-hadrian.nix:178]: But it might at some point. Having it in scope like the others is less confusing.

**Comment by sternenseemann** [comment]
Haven't forgotten this, a bit busy atm, I'll take care of it soon.

**Review by sternenseemann — APPROVED** [review]
`finalAttrs` seems to be a little bit slower, so it seems to be a good idea to avoid it for GHC with haskell packages eval being slow already.

One thing that we also could add is a `ghcPatches` argument (to ghc and hadrian), but if you don't need it (yet) we can do it at another time. LMK what you want to do.

**Comment by 9999years** [comment]
I think this PR gives a nice minimal footprint for this feature. Let's merge this, we can always revisit in the future.

**Pull request merged**